### PR TITLE
Fix typing and other misc cleanup

### PR DESCRIPTION
**Pull request by arcondello**
Also fixed alphabetization and added a few tests and argument checks.

### CODE REVIEW

File: dwave/optimization/model.pyi [hunk 1]
```diff
@@ -12,29 +12,166 @@
...
+        upper_bound: typing.Optional[int] = None,
+        ) -> IntegerVariable: ...
+
+    def into_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        *,
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ): ...
```
Comment (JoelPasvolsky) [inline dwave/optimization/model.pyi:69]: should this have a typing `->` typing.BinaryIO: ...` here?

File: dwave/optimization/model.pyi [hunk 2]
```diff
@@ -12,29 +12,166 @@
...
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ) -> typing.BinaryIO: ...
 
-class ListVariable(ArraySymbol):
-    def set_state(self, index: int, state: numpy.typing.ArrayLike) -> None: ...
+    # networkx might not be installed, so we just say we return an object.
+    def to_networkx(self) -> object: ...
 
+    def unlock(self): ...
```
Comment (JoelPasvolsky) [inline dwave/optimization/model.pyi:106]: Is there a typing for an exit form contextmanager? (equivalent to `-> typing.ContextManager`)

File: dwave/optimization/model.pyi [hunk 3]
```diff
@@ -12,29 +12,166 @@
...
+        check_header: bool = True,
+        ) -> Model: ...
+
+    def from_future(self, future: object, result_hook: typing.Callable): ...
+    def initialize(self): ...
+
+    def into_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        ): ...
```
Comment (JoelPasvolsky) [inline dwave/optimization/model.pyi:128]: typing `-> typing.BinaryIO: ...` here?

File: dwave/optimization/model.pyi [hunk 4]
```diff
@@ -12,29 +12,166 @@
...
+    def size(self) -> int: ...
+    def to_file(self) -> typing.BinaryIO: ...
+
+
+class Symbol:
+    def __init__(self, *args, **kwargs) -> typing.NoReturn: ...
+    def equals(self, other: Symbol) -> bool: ...
+    def expired(self) -> bool: ...
+    def has_state(self, index: int = 0) -> bool: ...
+    def iter_predecessors(self) -> typing.Iterator[Symbol]: ...
```
Comment (JoelPasvolsky) [inline dwave/optimization/model.pyi:141]: Why is this an iterator? The N successors have a single predecessor

File: dwave/optimization/model.pyi [hunk 5]
```diff
@@ -12,29 +12,166 @@
...
+    def iter_successors(self) -> typing.Iterator[Symbol]: ...
+    def maybe_equals(self, other: Symbol) -> int: ...
```
Comment (JoelPasvolsky) [inline dwave/optimization/model.pyi:143]: `int` or `bool`?

File: dwave/optimization/symbols.pyx [hunk 6]
```diff
@@ -1514,31 +1487,38 @@ cdef class ListVariable(ArraySymbol):
...
+            lb=self.lower_bound(),
+            ub=self.upper_bound(),
             )
 
         encoder = json.JSONEncoder(separators=(',', ':'))
 
         zf.writestr(directory + "shape.json", encoder.encode(shape_info))
 
+    def lower_bound(self):
+        """The lowest value the integer(s) can take (inclusive)."""
```
Comment (JoelPasvolsky) [inline dwave/optimization/symbols.pyx:1510]:
```suggestion
        """The lowest value allowed for the integer symbol."""
```

File: dwave/optimization/symbols.pyx [hunk 7]
```diff
@@ -1514,31 +1487,38 @@ cdef class ListVariable(ArraySymbol):
...
+        return int(self.ptr.lower_bound())
+
     def set_state(self, Py_ssize_t index, state):
-        """Set the state of the list node.
+        """Set the state of the integer node.
 
-        The given state must be a permuation of ``range(len(state))``.
+        The given state must be integer array of the integer node shape.
```
Comment (JoelPasvolsky) [inline dwave/optimization/symbols.pyx:1516]:
```suggestion
        The given state must be an integer array with the symbol's shape.
```

File: dwave/optimization/symbols.pyx [hunk 8]
```diff
@@ -1554,40 +1534,71 @@ cdef class ListVariable(ArraySymbol):
         # The validity of the state is checked in C++
         self.ptr.initialize_state(self.model.states._states[index], move(items))
 
-    # An observing pointer to the C++ ListNode
-    cdef cppListNode* ptr
+    def upper_bound(self):
+        """The highest value the integer(s) can take (inclusive)."""
```
Comment (JoelPasvolsky) [inline dwave/optimization/symbols.pyx:1538]:
```suggestion
        """The highest value allowed for the integer symbol."""
```

File: dwave/optimization/symbols.pyx [hunk 9]
```diff
@@ -1554,40 +1534,71 @@ cdef class ListVariable(ArraySymbol):
...
+        >>> j = model.integer(25, lower_bound=-100)
+        >>> k = i <= j
+        >>> type(k)
+        <class 'dwave.optimization.symbols.LessEqual'>
     """
-    def __init__(self, Model model, shape=None, lower_bound=None, upper_bound=None):
-        cdef vector[Py_ssize_t] vshape = _as_cppshape(tuple() if shape is None else shape )
+    def __init__(self, ArraySymbol lhs, ArraySymbol rhs):
+        if lhs.model is not rhs.model:
+            raise ValueError("lhs and rhs do not share the same underlying model")
```
Comment (JoelPasvolsky) [inline dwave/optimization/symbols.pyx:1561]:
```suggestion
            raise ValueError("left and right sides do not share the same underlying model")
```

**Review by JoelPasvolsky — APPROVED** [review]
Not familiar with all that's going on in here but what I understand looks good.

File: dwave/optimization/model.pyi (hunk 1 above)
Comment (arcondello) [reply dwave/optimization/model.pyi:69]: `to_file()` returns a file, but `into_file()` does not.

File: dwave/optimization/model.pyi (hunk 2 above)
Comment (arcondello) [reply dwave/optimization/model.pyi:106]: `unlock()` is called by the context manager, it doesn't return one. We do `lock(self) -> typing.ContextManager` which covers enter/exit.

File: dwave/optimization/model.pyi (hunk 3 above)
Comment (arcondello) [reply dwave/optimization/model.pyi:128]: See https://github.com/dwavesystems/dwave-optimization/pull/20/files#r1637072820

File: dwave/optimization/model.pyi (hunk 4 above)
Comment (arcondello) [reply dwave/optimization/model.pyi:141]: it's always `for pred in symbol.iter_predecessors(): ...`. I.e. an iterator.

File: dwave/optimization/model.pyi (hunk 5 above)
Comment (arcondello) [reply dwave/optimization/model.pyi:143]: `maybe_equals` returns an `int`. It's truthy, but `2` is also an allowed value. I could do `typing.Literal[0, 1, 2]` I suppose, but that feels like overkill.

File: dwave/optimization/symbols.pyx (hunk 8 above)
Comment (arcondello) [reply dwave/optimization/symbols.pyx]: Haha, this is your docstring! Or at least not mine. But sure.

File: dwave/optimization/symbols.pyx (hunk 9 above)
Comment (arcondello) [reply dwave/optimization/symbols.pyx:1561]: `lhs` and `rhs` are the parameter names.

File: dwave/optimization/model.pyi (hunk 1 above)
Comment (JoelPasvolsky) [reply dwave/optimization/model.pyi:69]: makes sense

**Pull request merged**